R/derive_vars_trt_start.R

```diff
@@ -0,0 +1,158 @@
+#'   `by_vars` parameter) the first observation (with respect to the order
+#'   specified for the `order` parameter) is included in the output dataset.
+#'
+#' @author Stefan Bundfuss, Teckla Akinyi
+#'
+#' @return The input dataset with start date variable added e.g. `TRTSDTM`
+#'
+#' @export
+#'
+#' @keywords adsl timing derivation
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("ex")
+#' data("dm")
+#'
+#' dm %>%
+#'   derive_vars_trt_start(dataset_ex = ex
+#'                         filter_ex =  (EXDOSE > 0 | (EXDOSE == 0 & str_detect(EXTRT, "PLACEBO"))) & nchar(EXSTDTC) >= 10, # nolint
+#'                         subject_keys = vars(STUDYID, USUBJID),
+#'                         new_var=TRTSDTM,
+#'                         dtc=EXSTDTC,
+#'                         date_imputation = "first",
+#'                         time_imputation="first",
+#'                         min_dates = NULL,
+#'                         max_dates = NULL,
+#'                         ord_vars = vars(EXSTDTC,EXSEQ),
+#'                         ord_filter="first") %>%
+#'   select(USUBJID, TRTSDTM)
+derive_vars_trt_start <- function(dataset,
+                                 dataset_ex,
+                                 filter_ex = NULL,
+                                 subject_keys,
+                                 new_var,
+                                 dtc,
+                                 date_imputation,
+                                 time_imputation,
+                                 min_dates = NULL,
+                                 max_dates = NULL,
+                                 ord_vars,
+                                 ord_filter) {
+    new_var <- assert_symbol(enquo(new_var))
+    dtc <- assert_symbol(enquo(dtc))
+    ord <- assert_symbol(enquo(ord))
+    assert_data_frame(dataset,subject_keys)
+    assert_data_frame(dataset_ex, required_vars = quo_c(subject_keys, ord_vars))
+    assert_character_scalar(ord_filter, values = c("first","last"))
+
+    filter_ex <- assert_filter_cond(enquo(filter_ex), optional = TRUE)
+
+    add <- dataset_ex %>%
+      mutate(imputed_dtc = convert_dtc_to_dtm(!!dtc,
+                                              date_imputation = date_imputation,
+                                              time_imputation = time_imputation,
+                                              min_dates = min_dates,
+                                              max_dates = max_dates)
+      )
+    if (!quo_is_null(filter_pre_timepoint)){
+      add %>% filter_if(filter_ex) %>%
+        filter_extreme(
+          order = vars(!!dtc, !!ord_vars),
+          by_vars = !!!subject_keys,
+          mode = ord_filter) %>%
+        mutate(!!new_var := imputed_dtc)
+    } else {
+      filter_extreme(
+        order = ord,
+        by_vars = !!!subject_keys,
+        mode = ord_filter) %>%
+        mutate(!!new_var := imputed_dtc)
+    }
+
+    left_join(dataset, add, by = vars2chr(subject_keys)) %>%
```

Comment:
At this stage the `add` datasets will contain all variables from `dataset_ex` and they will be added to the output dataset. Need to select only the relevant variables.